### PR TITLE
fix(core): parse upload headers correctly

### PR DESCRIPTION
**Pull request by kptkin**
Description
-----------

- Fixes WB-17203

What does the PR do?

- This PR fixes the file upload to pass headers as part of the task
- This PR fixes the logic that parses the header inputs, some headers might have `:` in the header values that we used to truncate, we this change we preserve them.
- Fix the download test to use fake end point and add another test for upload

Testing
-------
How was this PR tested?

### CODE REVIEW

File: core/internal/filetransfer/file_transfer_default.go [hunk 1]
```diff
@@ -63,14 +63,12 @@ func (ft *DefaultFileTransfer) Upload(task *Task) error {
 		return err
 	}
 	for _, header := range task.Headers {
-		parts := strings.Split(header, ":")
+		parts := strings.SplitN(header, ":", 2)
```
Comment (dmitryduev) [inline core/internal/filetransfer/file_transfer_default.go:66]: could you add a unit test for this please? (say with a legit-looking aws header, and something else)

**Review by dmitryduev — APPROVED** [review]
🙏

File: core/internal/filetransfer/file_transfer_default.go (hunk 1 above)
Comment (kptkin) [reply core/internal/filetransfer/file_transfer_default.go:66]: Sure... or is not done yet...

**Comment by estellazx** [comment]
I'm debugging but its failing for smaller artifacts right now but seems to be an error for larger artifacts. My code to reproduce error:

```
import wandb
import os
import logging
from pathlib import Path
from datetime import timedelta
import time

os.environ["WANDB_BASE_URL"] = "https://staging-aws.wandb.io"

logging.basicConfig(level=logging.INFO)
LOGGER = logging.getLogger()

def create_dummy_checkpoint(size_in_mb, file_path="model.ckpt"):
    LOGGER.info(
        f"Creating a dummy checkpoint file of size {size_in_mb}MB with random data"
    )
    # Generate random data to fill the file
    with open(file_path, mode="wb") as f:
        f.truncate(size_in_mb * 1024**2)


class ModelCheckpoint:
    def __init__(self, file_path):
        self._last_checkpoint_saved = file_path


class YourModel:
    def __init__(self, recovery_model_ttl_days=30, log_latest_model=True):
        self.recovery_model_ttl_days = recovery_model_ttl_days
        self.log_latest_model = log_latest_model

    def after_save_checkpoint(self, checkpoint_callback: ModelCheckpoint, run):
        if self.log_latest_model:
            self._log_latest_model_checkpoint(
                checkpoint_callback=checkpoint_callback, run=run
            )

    def _log_latest_model_checkpoint(self, checkpoint_callback: ModelCheckpoint, run):
        latest_ckpt_path = checkpoint_callback._last_checkpoint_saved
        artifact_name = f"recovery-model-{wandb.run.id}"

        metadata = {"original_filename": Path(latest_ckpt_path).name}
        new_artifact = wandb.Artifact(
            name=artifact_name, type="model", metadata=metadata
        )
        new_artifact.add_file(latest_ckpt_path, name="model.ckpt")

        LOGGER.info(f"Uploading {latest_ckpt_path} as artifact {artifact_name}...")
        tic = time.perf_counter()
        run.log_artifact(new_artifact, aliases=["latest"]).wait()
        toc = time.perf_counter()
        LOGGER.info(
            f"Uploaded {latest_ckpt_path} as artifact {artifact_name} in {toc - tic:0.4f} seconds"
        )

run = wandb.init(entity="estella-xin", project="test")

model = YourModel()

# Log small artifact, this doesn't fail
create_dummy_checkpoint(10, "small_model.ckpt")
small_checkpoint_callback = ModelCheckpoint("small_model.ckpt")
model.after_save_checkpoint(small_checkpoint_callback, run)

# Log medium artifact. FAILS
create_dummy_checkpoint(1024, "medium_model.ckpt")
medium_checkpoint_callback = ModelCheckpoint("medium_model.ckpt")
model.after_save_checkpoint(medium_checkpoint_callback, run)

run.finish()

```

**Comment by kptkin** [comment]
> I'm debugging but its failing for smaller artifacts right now but seems to be an error for larger artifacts. My code to reproduce error:
> 
> ```
> import wandb
> import os
> import logging
> from pathlib import Path
> from datetime import timedelta
> import time
> 
> os.environ["WANDB_BASE_URL"] = "https://staging-aws.wandb.io"
> 
> logging.basicConfig(level=logging.INFO)
> LOGGER = logging.getLogger()
> 
> def create_dummy_checkpoint(size_in_mb, file_path="model.ckpt"):
>     LOGGER.info(
>         f"Creating a dummy checkpoint file of size {size_in_mb}MB with random data"
>     )
>     # Generate random data to fill the file
>     with open(file_path, mode="wb") as f:
>         f.truncate(size_in_mb * 1024**2)
> 
> 
> class ModelCheckpoint:
>     def __init__(self, file_path):
>         self._last_checkpoint_saved = file_path
> 
> 
> class YourModel:
>     def __init__(self, recovery_model_ttl_days=30, log_latest_model=True):
>         self.recovery_model_ttl_days = recovery_model_ttl_days
>         self.log_latest_model = log_latest_model
> 
>     def after_save_checkpoint(self, checkpoint_callback: ModelCheckpoint, run):
>         if self.log_latest_model:
>             self._log_latest_model_checkpoint(
>                 checkpoint_callback=checkpoint_callback, run=run
>             )
> 
>     def _log_latest_model_checkpoint(self, checkpoint_callback: ModelCheckpoint, run):
>         latest_ckpt_path = checkpoint_callback._last_checkpoint_saved
>         artifact_name = f"recovery-model-{wandb.run.id}"
> 
>         metadata = {"original_filename": Path(latest_ckpt_path).name}
>         new_artifact = wandb.Artifact(
>             name=artifact_name, type="model", metadata=metadata
>         )
>         new_artifact.add_file(latest_ckpt_path, name="model.ckpt")
> 
>         LOGGER.info(f"Uploading {latest_ckpt_path} as artifact {artifact_name}...")
>         tic = time.perf_counter()
>         run.log_artifact(new_artifact, aliases=["latest"]).wait()
>         toc = time.perf_counter()
>         LOGGER.info(
>             f"Uploaded {latest_ckpt_path} as artifact {artifact_name} in {toc - tic:0.4f} seconds"
>         )
> 
> run = wandb.init(entity="estella-xin", project="test")
> 
> model = YourModel()
> 
> # Log small artifact, this doesn't fail
> create_dummy_checkpoint(10, "small_model.ckpt")
> small_checkpoint_callback = ModelCheckpoint("small_model.ckpt")
> model.after_save_checkpoint(small_checkpoint_callback, run)
> 
> # Log medium artifact. FAILS
> create_dummy_checkpoint(1024, "medium_model.ckpt")
> medium_checkpoint_callback = ModelCheckpoint("medium_model.ckpt")
> model.after_save_checkpoint(medium_checkpoint_callback, run)
> 
> run.finish()
> ```

i'm not sure it is the same things, since i tested with regular files and i see that is uploaded correctly. I'm guessing it's probably something artifact specific

**Comment by kptkin** [comment]
> > I'm debugging but its failing for smaller artifacts right now but seems to be an error for larger artifacts. My code to reproduce error:
> > ```
> > import wandb
> > import os
> > import logging
> > from pathlib import Path
> > from datetime import timedelta
> > import time
> > 
> > os.environ["WANDB_BASE_URL"] = "https://staging-aws.wandb.io"
> > 
> > logging.basicConfig(level=logging.INFO)
> > LOGGER = logging.getLogger()
> > 
> > def create_dummy_checkpoint(size_in_mb, file_path="model.ckpt"):
> >     LOGGER.info(
> >         f"Creating a dummy checkpoint file of size {size_in_mb}MB with random data"
> >     )
> >     # Generate random data to fill the file
> >     with open(file_path, mode="wb") as f:
> >         f.truncate(size_in_mb * 1024**2)
> > 
> > 
> > class ModelCheckpoint:
> >     def __init__(self, file_path):
> >         self._last_checkpoint_saved = file_path
> > 
> > 
> > class YourModel:
> >     def __init__(self, recovery_model_ttl_days=30, log_latest_model=True):
> >         self.recovery_model_ttl_days = recovery_model_ttl_days
> >         self.log_latest_model = log_latest_model
> > 
> >     def after_save_checkpoint(self, checkpoint_callback: ModelCheckpoint, run):
> >         if self.log_latest_model:
> >             self._log_latest_model_checkpoint(
> >                 checkpoint_callback=checkpoint_callback, run=run
> >             )
> > 
> >     def _log_latest_model_checkpoint(self, checkpoint_callback: ModelCheckpoint, run):
> >         latest_ckpt_path = checkpoint_callback._last_checkpoint_saved
> >         artifact_name = f"recovery-model-{wandb.run.id}"
> > 
> >         metadata = {"original_filename": Path(latest_ckpt_path).name}
> >         new_artifact = wandb.Artifact(
> >             name=artifact_name, type="model", metadata=metadata
> >         )
> >         new_artifact.add_file(latest_ckpt_path, name="model.ckpt")
> > 
> >         LOGGER.info(f"Uploading {latest_ckpt_path} as artifact {artifact_name}...")
> >         tic = time.perf_counter()
> >         run.log_artifact(new_artifact, aliases=["latest"]).wait()
> >         toc = time.perf_counter()
> >         LOGGER.info(
> >             f"Uploaded {latest_ckpt_path} as artifact {artifact_name} in {toc - tic:0.4f} seconds"
> >         )
> > 
> > run = wandb.init(entity="estella-xin", project="test")
> > 
> > model = YourModel()
> > 
> > # Log small artifact, this doesn't fail
> > create_dummy_checkpoint(10, "small_model.ckpt")
> > small_checkpoint_callback = ModelCheckpoint("small_model.ckpt")
> > model.after_save_checkpoint(small_checkpoint_callback, run)
> > 
> > # Log medium artifact. FAILS
> > create_dummy_checkpoint(1024, "medium_model.ckpt")
> > medium_checkpoint_callback = ModelCheckpoint("medium_model.ckpt")
> > model.after_save_checkpoint(medium_checkpoint_callback, run)
> > 
> > run.finish()
> > ```
> 
> i'm not sure it is the same things, since i tested with regular files and i see that is uploaded correctly. I'm guessing it's probably something artifact specific

I tried to upload the same sized file using the regular file upload and it worked. So i'm guessing it is something in the artifact logic specifically
```
run = wandb.init(project="test", settings=wandb.Settings(_disable_stats=True))
create_dummy_checkpoint(1024, "medium_model.ckpt")
run.save("medium_model.ckpt", base_path=".", policy="now")
run.finish()
```

**Comment by estellazx** [comment]
> > > I'm debugging but its failing for smaller artifacts right now but seems to be an error for larger artifacts. My code to reproduce error:
> > > ```
> > > import wandb
> > > import os
> > > import logging
> > > from pathlib import Path
> > > from datetime import timedelta
> > > import time
> > > 
> > > os.environ["WANDB_BASE_URL"] = "https://staging-aws.wandb.io"
> > > 
> > > logging.basicConfig(level=logging.INFO)
> > > LOGGER = logging.getLogger()
> > > 
> > > def create_dummy_checkpoint(size_in_mb, file_path="model.ckpt"):
> > >     LOGGER.info(
> > >         f"Creating a dummy checkpoint file of size {size_in_mb}MB with random data"
> > >     )
> > >     # Generate random data to fill the file
> > >     with open(file_path, mode="wb") as f:
> > >         f.truncate(size_in_mb * 1024**2)
> > > 
> > > 
> > > class ModelCheckpoint:
> > >     def __init__(self, file_path):
> > >         self._last_checkpoint_saved = file_path
> > > 
> > > 
> > > class YourModel:
> > >     def __init__(self, recovery_model_ttl_days=30, log_latest_model=True):
> > >         self.recovery_model_ttl_days = recovery_model_ttl_days
> > >         self.log_latest_model = log_latest_model
> > > 
> > >     def after_save_checkpoint(self, checkpoint_callback: ModelCheckpoint, run):
> > >         if self.log_latest_model:
> > >             self._log_latest_model_checkpoint(
> > >                 checkpoint_callback=checkpoint_callback, run=run
> > >             )
> > > 
> > >     def _log_latest_model_checkpoint(self, checkpoint_callback: ModelCheckpoint, run):
> > >         latest_ckpt_path = checkpoint_callback._last_checkpoint_saved
> > >         artifact_name = f"recovery-model-{wandb.run.id}"
> > > 
> > >         metadata = {"original_filename": Path(latest_ckpt_path).name}
> > >         new_artifact = wandb.Artifact(
> > >             name=artifact_name, type="model", metadata=metadata
> > >         )
> > >         new_artifact.add_file(latest_ckpt_path, name="model.ckpt")
> > > 
> > >         LOGGER.info(f"Uploading {latest_ckpt_path} as artifact {artifact_name}...")
> > >         tic = time.perf_counter()
> > >         run.log_artifact(new_artifact, aliases=["latest"]).wait()
> > >         toc = time.perf_counter()
> > >         LOGGER.info(
> > >             f"Uploaded {latest_ckpt_path} as artifact {artifact_name} in {toc - tic:0.4f} seconds"
> > >         )
> > > 
> > > run = wandb.init(entity="estella-xin", project="test")
> > > 
> > > model = YourModel()
> > > 
> > > # Log small artifact, this doesn't fail
> > > create_dummy_checkpoint(10, "small_model.ckpt")
> > > small_checkpoint_callback = ModelCheckpoint("small_model.ckpt")
> > > model.after_save_checkpoint(small_checkpoint_callback, run)
> > > 
> > > # Log medium artifact. FAILS
> > > create_dummy_checkpoint(1024, "medium_model.ckpt")
> > > medium_checkpoint_callback = ModelCheckpoint("medium_model.ckpt")
> > > model.after_save_checkpoint(medium_checkpoint_callback, run)
> > > 
> > > run.finish()
> > > ```
> > 
> > 
> > i'm not sure it is the same things, since i tested with regular files and i see that is uploaded correctly. I'm guessing it's probably something artifact specific
> 
> I tried to upload the same sized file using the regular file upload and it worked. So i'm guessing it is something in the artifact logic specifically
> 
> ```
> run = wandb.init(project="test", settings=wandb.Settings(_disable_stats=True))
> create_dummy_checkpoint(1024, "medium_model.ckpt")
> run.save("medium_model.ckpt", base_path=".", policy="now")
> run.finish()
> ```

@kptkin I looked at the upload URL and the headers look consistent and also I don't think it has anything to do with this PR, will add separate ticket

**Review by estellazx — APPROVED** [review]
tested locally and worked for smaller files (need to debug issue with larger files)

File: core/pkg/server/sender.go [hunk 2]
```diff
@@ -1058,9 +1059,9 @@ func (s *Sender) sendFile(file *service.FilesItem) {
 				s.fwdChan <- record
 			},
 		)
-		task.AddCompletionCallback(s.fileTransferManager.FileStreamCallback())
-		task.AddCompletionCallback(
+		task.SetCompletionCallback(
 			func(*filetransfer.Task) {
+				s.fileTransferManager.FileStreamCallback()
```
Comment (timoffex) [inline core/pkg/server/sender.go:1064]: This function seems to return a callback; should the return value be invoked?

```go
s.fileTransferManager.FileStreamCallback()()
```

though if that's the case, it's definitely a little error prone

**Review by timoffex — APPROVED** [review]

**Review by timoffex — APPROVED** [review]

**Pull request merged**